### ISSUE TITLE
Proposal: Clouformation wait for signal 

### ISSUE THREAD

**Issue by gianrubio** [coreos/kube-aws#49]
In the current cloudformation template, kube-aws use PauseTime flag to wait for create/update resources.  

```
    UpdatePolicy" : {
        "AutoScalingRollingUpdate" : {
          "MinInstancesInService" : "{{.ControllerCount}}",
          "MaxBatchSize" : "1",
          "PauseTime" : "PT2M"
        }
      }
```

My proposal is to use cfn-signal to control the resources [1][2] and only send the signal after the last service was started. I can implement this proposal and I'm here to listen for suggestions.

```
[Unit]
Description=Cloudformation Signal Ready
After=docker.service
Requires=docker.service
After=etcd2.service
Requires=etcd2.service

[Install]
WantedBy=multi-user.target

[Service]
Type=oneshot
EnvironmentFile=/etc/environment
ExecStart=/bin/bash -c '\
eval $(docker run crewjam/ec2cluster); \
docker run --rm crewjam/awscli cfn-signal \
    --resource MasterAutoscale --stack $TAG_AWS_CLOUDFORMATION_STACK_ID \
    --region $REGION; \
'
```

[1] https://aws.amazon.com/blogs/devops/use-a-creationpolicy-to-wait-for-on-instance-configurations/
[2] http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-creationpolicy.html

**Comment by pieterlange** [comment]
Big fan of this idea but i'm seeing some issues here:
  - Running containers straight from docker hub outside of the CoreOS management sphere.
  - We obviously would need to do this for both the `AutoScaleController` and `AutoScaleWorker` groups.

We don't actually need `crewjam/ec2cluster` here as all we need to set is `$TAG_AWS_CLOUDFORMATION_STACK_ID` and `$REGION`. We could set both environment variables using the feature described here: https://github.com/coreos/kube-aws/issues/29

There is an image available from CoreOS to replace the `crewjam/awscli` docker command in order to run `cfn-signal`: https://quay.io/repository/coreos/awscli but it could use an update (@colhom).

(Full disclosure: @gianrubio is a colleague :-) )

**Comment by iwarp** [comment]
Also big fan of this, In the UpdatePolicy you will have to include WaitOnResourceSignals=true and set the PauseTime higher in order to give time for the system to come up, otherwise cloudformation will rollback the deploy.

**Comment by mumoshu** [comment]
@gianrubio Also big fan of this!

I'd like to add that, if we're going to apply this to anything other than `cloud-config-etcd`, we have to workaround the known systemd issue that prevents services from restarting when they're required via `Requires`.

// It'd been very tricky to handle so please let me continue with longer explanation..

Well, in such case, I believe we should at-least replace `Requires` with `Wants` and optionally wait for `want`ed services to be up with `Restart=on-failure` +  `ExecStartPre=/usr/bin/systemctl is-active <wanted service>`.

You can check out https://github.com/coreos/kube-aws/commit/dc7aafa2bdcfced5575e85b6e95609595811dda3 to see how I've been doing this.

Also note that `Restart=on-failure` doesn't work with `Type=oneshot`. So in many cases we'll probably end up translating it to `Type=simple`.

I'm saying "anything other than `cloud-config-etcd`" here because I believe etcd nodes don't have the startup instability issue like worker/controller thus no fixes like that applied to worker/controller.
So you can probably apply your implementation as is to `cloud-config-etcd`, but not to `cloud-config-worker` and `cloud-config-controller`.

See
* https://github.com/coreos/coreos-kubernetes/pull/697
* https://github.com/coreos/coreos-kubernetes/issues/675
* and another issues referenced from these
for related discussions.

**Comment by pieterlange** [comment]
We'll make sure to do proper testing on this taking special care of the systemd trickyness. I believe @gianrubio simply copied an excerpt from crewjam-etcd for demonstration purposes only.

This wait-for-signal PR wouldn't apply to the kube-aws embedded etcd nodes as they're:
  * locked resources
  * not part of an autoscalinggroup but in fact separately booted instances (fixed)

**Comment by mumoshu** [comment]
Ah, thanks for the correction!
I'd been forgetting about those two things and blindly mentioned about `cloud-config-etcd`.
Certainly, it won't apply to etcd nodes. Only controllers/workers which are updatable would be addressed in the PR.

**Comment by gianrubio** [comment]
Sorry if my first idea was not so clear, @pieterlange already detailed.

@mumoshu I think we should always use signals for cloud formation bootstrapping, it's the only way to guarantee that the cluster was successfully created/updated, we could apply this for etcd (create) and for workers/controlleres (create/update)

**Comment by mumoshu** [comment]
@gianrubio Thanks for your clarification. I totally agree to you.
The distinction between create and update is nice 👍 

Let me also add that, from an user's perspective, cfn signals for updates would be more critical than for creations, because there's already an possible issue related to it which incurs downtime in certain case https://github.com/coreos/kube-aws/issues/50

Implementing signals for creations would also be nice to have because it will allow me to snip the `wait until the k8s api server become responsible` part of my automations script like [this](https://github.com/coreos/kube-aws/blob/master/e2e/run#L98).

**Comment by mumoshu** [comment]
@gianrubio Thanks for the former pull request addressing this for controller and worker nodes 👍 
Are you willing to work on the same for etcd nodes, too?
If so, I'll leave this issue open.

**Comment by gianrubio** [comment]
@mumoshu I guess we could wait to resolution of #27, what do you think?

**Comment by mumoshu** [comment]
Sounds good but @pieterlange, are you going to address this in #27?

**Comment by pieterlange** [comment]
Yes, this will be addressed when i finally make time to do #27

**Comment by mumoshu** [comment]
Can I close this once #332 is merged?

**Comment by gianrubio** [comment]
@mumoshu sure!

**Comment by mumoshu** [comment]
Thanks for the confirmation @gianrubio!
I'm closing as #332 is merged.
